examples/roads2png.rs

```diff
@@ -166,46 +119,34 @@ fn roads<'a>(
     let ways = archive.ways();
     let tags_index = archive.tags_index();
     let tags = archive.tags();
-    // we checked strings in the beginning, it's safe to do 0-overhead retrieval.
-    let strings = unsafe { str::from_utf8_unchecked(archive.stringtable()) };
+    let strings = archive.stringtable();
 
     ways.iter()
         .tuple_windows()
         .filter(move |(way, next_way)| way_filter(&way, &next_way, &tags_index, &tags, strings))
 }
 
 fn render(archive: &osmflat::Osm, width: u32) -> Image {
-    str::from_utf8(archive.stringtable()).expect("stringtable contains invalid utf8 characters");
-
     // compute extent
-    let mut coords =
-        roads(archive).flat_map(|(way, next_way)| way_coords(archive, &way, &next_way));
-    let first_coord = coords.next().expect("no roads found");
-    let (min, max) = coords.fold((first_coord, first_coord), |(min, max), coord| {
-        (min.min(coord), max.max(coord))
-    });
+    let coords = roads(archive).flat_map(|(way, next_way)| way_nodes(archive, &way, &next_way));
+    let (min, max) = compute_bounds(coords);
 
     // compute ratio and height
-    let ratio = 360. / 180. * (max.lat - min.lat) / (max.lon - min.lon);
+    let ratio = 360. / 180. * (max.0 - min.0) as f64 / (max.1 - min.1) as f64;
     let height = (width as f64 * ratio) as u32;
 
     // create world -> raster transformation
-    let t = MapTransform::new(width - 1, height - 1, min, max);
+    let t = map_transform(width - 1, height - 1, min, max);
 
     // draw
     let mut image = Image::new(width, height);
 
-    let lines = roads(archive).flat_map(|(way, next_way)| {
-        way_coords(archive, &way, &next_way)
-            .map(|coord| t.transform(coord))
-            .tuple_windows()
-    });
+    let line_segments = roads(archive)
+        .flat_map(|(way, next_way)| way_nodes(archive, &way, &next_way).map(t).tuple_windows());
 
-    for (from, to) in lines {
-        for (x, y) in Bresenham::new(from, to) {
-            image.set_black(x as u32, y as u32);
-        }
-    }
+    line_segments
+        .flat_map(|(from, to)| Bresenham::new(from, to))
+        .for_each(|(x, y)| image.set_black(x as u32, y as u32));
```

Comment:
Sometime I wonder, what is idiomatic, just a for loop or this for each.

Reply:
No idea yet. I have a slight preference for loops at the moment, and definitely do not like long iterator chains without giving intermediate things names, but otherwise not yet sure.

Reply:
Same here, I try to use loops instead of for_each.

Reply:
Reverted this